Attachments/adding-attachments.md

```diff
@@ -0,0 +1,251 @@
+
+Replace `USERNAME `, `PASSWORD` and `INSTALLATION_NAME` with valid values for your installation of Lime Field.
+
+The `entityHref` should point out the entity to which we add attachments.
+
+```javascript
+const username = 'USERNAME';
+const password = 'PASSWORD';
+const credentials = window.btoa(`${username}:${password}`);
+const baseurl = 'https://INSTALLATION_NAME.remotex.net/api';
+const entityHref = 'entity/id';
+
+function requestHeaders(options = {}) {
+  const result = new Headers({
+    'Accept': 'application/json',
+    'Authorization': `Basic ${credentials}`,
+    'Content-Type': 'application/json'
+  });
+
+  for (let header in options) {
+    result.set(header, options[header]);
+  }
+
+  return result;
+}
+
+function requestOptions(options = {}, headers = {}) {
+  return Object.assign({
+    method: 'POST',
+    cache: 'no-cache',
+    credentials: 'include',
+    headers: requestHeaders(headers)
+  }, options);
+}
+```
+
+### 1. Creating a `File` entity
+
+```javascript
+async function createFileEntity(file) {
+  const entity = {
+    Title: file.name,
+    FileName: file.name,
+    ContentType: file.type,
+    Length: file.size
+  };
+
+  return fetch(`${baseurl}/files`, requestOptions({
+    body: JSON.stringify(entity)
+  }))
+  .then(async response => ({
+    entity: JSON.parse(await response.text()),
+    file: file
+  }));
+}
+```
+
+### 2. Uploading the file content
+
+```javascript
+async function uploadFileContent(data) {
+  return fetch(`${baseurl}/${data.entity.Content.Href}`, requestOptions({
+    method: 'PUT',
+    body: data.file
+  }, {
+    'Accept': '*/*',
+    'Content-Type': data.file.type
+  }));
+}
+```
+
+### 3. Adding the files as an attachments
+
+__Note__, in this example we allow for multiple files uploads. But it works for a single file as well.
+
+```javascript
+async function addAttachmentsToEntity(attachments) {
+  const commands = attachments.map(attachment => ({
+    Name: 'AddAttachmentToEntity',
+    Target: entityHref,
+    Parameters: [{
+      Name: 'File', Values: [attachment.entity.Href]
+    }]
+  }));
+
+  return fetch(`${baseurl}/commands`, requestOptions({
+    body: JSON.stringify({Commands: commands})
+  }));
+}
+```
+
+### Bootstraping the code to a view
+
+Bind `openFilesManager` to a `onclick` event on some button on your web page.
+
+__Note__, we do not await content upload requests. They can be executed in parallel.
+
+```javascript
+function isFileToLarge(file) {
+  return file && file.size > 10485760;
+}
+
+function openFilesManager() {
+  const input = document.createElement('input');
+  input.setAttribute('type', 'file');
+  input.setAttribute('accept', '*');
+  input.setAttribute('multiple', true);
+
+  input.addEventListener('change', event => {
+    const files = [...event.target.files];
+
+    if (files.some(isFileToLarge)) {
+      throw 'One of the files is to large';
```

Comment:
Maybe change to something like: `One or more of the files are too large`?

Reply:
This actually aborts on the first too large file. As this code snippet is en example, the adaptors of it are free to add their own validation, messages and what ever custom logic they need. 

This conditional check is included in the example because we do have a file size limit on our installations. At least used to have, not sure now with S3 storage, but pretty sure we did not remove it.